frame/nomination-pools/src/lib.rs

```diff
@@ -1799,67 +1802,117 @@ pub mod pallet {
 		}
 
 		#[pallet::weight(T::WeightInfo::nominate(validators.len() as u32))]
+		#[transactional]
 		pub fn nominate(
 			origin: OriginFor<T>,
 			pool_id: PoolId,
 			validators: Vec<T::AccountId>,
 		) -> DispatchResult {
-			let who = ensure_signed(origin)?;
-			let bonded_pool = BondedPool::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
-			ensure!(bonded_pool.can_nominate(&who), Error::<T>::NotNominator);
-			T::StakingInterface::nominate(bonded_pool.bonded_account(), validators)?;
+			let bonded_account = Self::ensure_signed_or_root(
+				origin,
+				|who| {
+					let bonded_pool =
+						BondedPool::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
+					ensure!(bonded_pool.can_nominate(&who), Error::<T>::NotNominator);
+					Ok(bonded_pool.bonded_account())
+				},
+				|| {
+					let bonded_pool =
+						BondedPool::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
+					Ok(bonded_pool.bonded_account())
+				},
+			)?;
+
+			T::StakingInterface::nominate(bonded_account, validators.clone())?;
+			Self::deposit_event(Event::<T>::Nominated { pool_id, validators });
+
 			Ok(())
 		}
 
+		/// Set the metadata of a pool.
+		///
+		/// Dispatch origin of this call must be signed by the root or state-toggler roles of the
+		/// pool, or be the (holy) root origin.
 		#[pallet::weight(T::WeightInfo::set_state())]
 		pub fn set_state(
 			origin: OriginFor<T>,
 			pool_id: PoolId,
 			state: PoolState,
 		) -> DispatchResult {
-			let who = ensure_signed(origin)?;
-			let mut bonded_pool = BondedPool::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
-			ensure!(bonded_pool.state != PoolState::Destroying, Error::<T>::CanNotChangeState);
-
-			if bonded_pool.can_toggle_state(&who) {
-				bonded_pool.set_state(state);
-			} else if bonded_pool.ok_to_be_open(Zero::zero()).is_err() &&
-				state == PoolState::Destroying
-			{
-				// If the pool has bad properties, then anyone can set it as destroying
-				bonded_pool.set_state(PoolState::Destroying);
-			} else {
-				Err(Error::<T>::CanNotChangeState)?;
-			}
+			let bonded_pool = Self::ensure_signed_or_root(
+				origin,
+				|who| {
+					let mut bonded_pool =
+						BondedPool::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
+					if bonded_pool.ok_to_be_open(Zero::zero()).is_err() &&
+						state == PoolState::Destroying
+					{
+						// if the pool is in a bad state, we force set it to `Destroying`..
+						bonded_pool.set_state(PoolState::Destroying);
+					} else {
+						// else we set it to whatever the call wanted.
+						ensure!(
+							bonded_pool.state != PoolState::Destroying &&
+								bonded_pool.can_toggle_state(&who),
+							Error::<T>::CanNotChangeState
+						);
+						bonded_pool.set_state(state);
+					}
+					Ok(bonded_pool)
+				},
+				|| {
```

Comment:
I'm noticing we are not checking `ok_to_be_open()` if root. Am I correct in assuming that governance's role here is to override these checks and force a state - such as re-opening a pool even if there are `OverflowRisk`s?